docs/src/references/mappers/c8y-mapper.md

```diff
@@ -1325,3 +1325,191 @@ te/device/main///cmd/log_upload/<cmd_id>
+    MQTT Broker ->> Tedge Mapper: Telemetry message to `te/device/nested1///m/temperature`
+    Tedge Mapper ->> Tedge Mapper: Auto-register the `nested1` device as immediate child device
+    MQTT Broker ->> Tedge Mapper: Deliver child registration message on `te/device/nested1//` `{ "@parent": "device/child1//" }`
+    Tedge Mapper ->> Tedge Mapper: Ignore the child registration message as the entity is already registered
+```
+
+Here a device that was meant to be registered as a nested child device got registered as immediate child device,
+which can't be undone/corrected when the actual registration message eventually arrives.
+
+#### Initial solution thoughts
+
+One obvious solution to avoid this problem is to force the users to disable auto-registration,
+when they intend to use explicit registration even for a single entity.
+But the downside with that approach is that we force them to use explicit registration for everything,
+even the entities for which auto-registration would have been perfectly fine,
+just because a few entities required some customizations via explicit registration.
+
+The other approach would be to resort to auto-registration only when
+an explicit registration message is not delivered within a predefined window(timeout).
+But, the same timing uncertainty issues described above, applies here as well:
+what if the explicit registration message is received after the auto-registration has already happened after the timeout?
+Crashing itself isn't gonna help here either, as the entity is already auto-registered.
+
+### Final Solution
+
+The ordering problems when the mapper is live and when it is restarted would be handled differently.
+
+#### Phase 1
+
+**Live mapper**
+
+When the mapper is live, if telemetry messages for a new entity is received before its registration message,
+they are simply dropped until that registration message is processed, with the assumption that,
+telemetry data updates will repeat eventually and hence will make up for the lost ones.
+But, all non-telemetry data like twin updates, capability messages etc are cached in-memory,
+and will be processed and registered in the entity store once the registration message eventually arrives.
+Keeping them cached in-memory is not a big deal, as they are eventually going to end-up in the in-memory entity store anyway.
+
+**On mapper restart**
+
+Sticking to the same solution of dropping telemetry messages on a mapper restart might result in loosing too much data,
+when the volume of queued telemetry messages are very high on the broker,
+increasing the probability of a lot of those telemetry messages getting delivered to the mapper first,
+before their respective registration messages are.
+
+To avoid so much data loss, on restart the mapper will selectively subscribe to the registration topics first,
+process all the retained registration messages (with a timeout), after which the regular message processing will start.
```

Comment:
If I understand correctly, the plan is to gather all the registration messages and to re-order them so parents can be registered before the child entities and to *rebuild* the state reached before the mapper stopped.

What about a persisted representation of the mapper entity store? This store will be updated on the fly (as described above under *live mapper* and persisted on disk after each new registration. On start, the mapper simply uses this restored entity store and updates it on the fly, if there are still new registrations.

Reply:
> What about a persisted representation of the mapper entity store?

Relying purely on the persisted entity store from the last runtime won't be enough, as there could be registration messages and other messages published when the mapper is down. On a restart, we'd want to avoid any races among those "previously unseen" messages as well. For that, a selective subscription to the registration topics on startup would be needed anyway.

We had considered this idea of "persisted entity store" for a different problem in the past: to distinguish already registered entities from any new ones registered when the mapper was down (to avoid resending all those already registered twin messages on mapper restart). So, this would still be needed, but unfortunately this alone won't solve the ordering problem that we're dealing with.

Reply:
The previous comment was made with the assumption that you're proposing it as an alternative to proposal 2 which is trying to avoid any message drops (even telemetry messages). But, since you clarified that it was proposed as an enhancement of proposal 1 itself, with its limitation around dropping some telemetry messages, the proposed enhancement is definitely easier to implement and we can go ahead with that.

> On a restart, we'd want to avoid any races among those "previously unseen" messages as well. For that, a selective subscription to the registration topics on startup would be needed anyway.

This was to minimize dropping as many telemetry messages as possible on restart (even the previously unseen telemetry messages delivered before their corresponding entity registration messages). But, if dropping some of those messages are acceptable (as we're okay dropping some telemetry messages during a live execution anyway), the proposal to use persistent entity store really simplifies the restart logic. It will also help in avoiding the "duplicate entity registration on restart" problem as well.

Reply:
So the proposal is to:

- persist the entity store, so on restart the mapper don't have to process from scratch all the registration messages, but only those published during down-time, limiting ordering issues
- persist in-memory registration messages that cannot be processed till a parent entity is registered
- (possibly) persist on-disk messages for unregistered entities till the registration is received (this would be a bit more involved as this might require a log per unregistered entity).

Is this correct?